src/meili/match_route.cc

```diff
@@ -224,30 +218,67 @@ std::vector<EdgeSegment> ConstructRoute(const MapMatcher& mapmatcher,
       }
 
       // we need to cut segments where a match result is marked as a break
-      new_segments.clear();
-      cut_segments(match_results, prev_idx, curr_idx, segments, new_segments);
-
-      // have to merge route's last segment and segments' first segment together if its not a
-      // discontinuity or a break
-      if (!prev_match->is_break_point && !route.empty() && !route.back().discontinuity &&
-          route.back().edgeid == new_segments.front().edgeid) {
-        // we modify the first segment of the new_segments accordingly to replace the previous
-        // one in the route.
-        new_segments.front().source = route.back().source;
-        // Prefer first_match_idx from previous segments but do not replace valid value with invalid.
-        if (route.back().first_match_idx != -1)
-          new_segments.front().first_match_idx = route.back().first_match_idx;
-        // Prefer last_match_idx from new segments but do not replace valid value with invalid.
-        if (new_segments.front().last_match_idx == -1)
-          new_segments.front().last_match_idx = route.back().last_match_idx;
-        route.pop_back();
+      // this loops through all interpolated points in between the stateful points (if any)
+      // and cuts segments and updates their first/last matched point idx's
+      auto first_segment = segments.begin();
+      for (int cut_begin_idx = prev_idx, cut_end_idx = prev_idx + 1; cut_end_idx <= curr_idx;
+           ++cut_end_idx) {
```

Comment:
the biggest change is this loop where we go through all `interpolated` points in between the matched points.